crypto/x509/v3_asid.c

```diff
@@ -300,7 +300,7 @@ static int ASIdentifierChoice_is_canonical(ASIdentifierChoice *choice)
             ASN1_INTEGER_to_BN(a_max, bn) == NULL ||
             !BN_add_word(bn, 1) ||
             (a_max_plus_one =
-             BN_to_ASN1_INTEGER(bn, a_max_plus_one)) == NULL) {
+             BN_to_ASN1_INTEGER(bn, old_max = a_max_plus_one)) == NULL) {
```

Comment:
I think the same error is here:
```
crypto/ec/ec_asn1.c:    ret->order = BN_to_ASN1_INTEGER(tmp, ret->order);
crypto/ec/ec_asn1.c:        ret->cofactor = BN_to_ASN1_INTEGER(tmp, ret->cofactor);
```

Reply:
I think you should do this
```C
        if(...
            !BN_add_word(bn, 1) ||
            (new_max =
              BN_to_ASN1_INTEGER(bn, a_max_plus_one)) == NULL) {
           goto done;
       }
       a_max_plus_one = new_max;
...
done:
     ASN1_INTEGER_free(a_max_plus_one);
     BN_free(bn);
     return ret;
```

Reply:
I'm adopting the suggested solution.

Reply:
It caused problems in the tests (both this and the EC equivalents), I've dropped the two CIDs from this PR: 1453631 and 1453626.

Reply:
These look to be false positives.  `BN_to_ASN1_INTEGER` calls down to `ASN1_STRING_set` which returns the `ASN1_STRING` string it was passed (unless it was passed NULL, in which case it allocates one).

Reply:
> crypto/ec/ec_asn1.c:    ret->order = BN_to_ASN1_INTEGER(tmp, ret->order);
> crypto/ec/ec_asn1.c:        ret->cofactor = BN_to_ASN1_INTEGER(tmp, ret->cofactor);

These aren't a problem because they appear to only ever be called with a NULL ECPARAMETERS argument, causing an allocation.

Reply:
That is not obvious to me, since ret can be passed in and why should order / cofactor
then be NULL?

https://github.com/openssl/openssl/blob/03e5668343078b963cc6544ad7270743de13e514/crypto/ec/ec_asn1.c#L453-L459

Reply:
The rest of OpenSSL only ever passes NULL in, it took some backtracking up call graphs to figure this out -- this is why Coverity didn't complain about these even though it seemed like it should have.

A fix similar to that in #9833 would be appropriate here, since it is a public facing API and we'll not know if NULL is always passed in or not.  This function is lacking documentation too.